alpenhorn/update.py

```diff
@@ -0,0 +1,466 @@
+            # fake it.
+            try:
+                link_path = os.path.join(node.root, req.file.acq.name, req.file.name)
+
+                # Check explicitly if link already exists as this and
+                # being unable to link will both raise OSError and get
+                # confused.
+                if os.path.exists(link_path):
+                    log.error('File %s already exists. Clean up manually.' % link_path)
+                    ret = -1
+                else:
+                    os.link(from_path, link_path)
+                    ret = 0
+                    md5sum = req.file.md5sum  # As we're linking the md5sum can't change. Skip the check here...
+
+            # If we couldn't just link the file, try copying it with rsync.
+            except OSError:
+                if util.command_available('rsync'):
+                    cmd = "rsync -%s %s %s" % (RSYNC_FLAG, from_path, to_dir)
+                    ret, stdout, stderr = util.run_command(cmd)
+
+                    md5sum = util.md5sum_file(to_file) if ret == 0 else None
+                else:
+                    log.warn("No commands available to complete this transfer.")
+                    ret = -1
+
+        # Check the return code...
+        if ret:
+            # If the copy didn't work, then the remote file may be corrupted.
+            log.error("Rsync failed. Marking source file suspect.")
+            ar.ArchiveFileCopy.update(has_file='M').where(
+                ar.ArchiveFileCopy.file == req.file,
+                ar.ArchiveFileCopy.node == req.node_from).execute()
+            continue
+        end_time = time.time()
+
+        # Check integrity.
+        if md5sum == req.file.md5sum:
+            size_mb = req.file.size_b / 2**20.0
+            trans_time = end_time - start_time
+            rate = size_mb / trans_time
+            log.info("Pull complete (md5sum correct). Transferred %.1f MB in %i "
+                     "seconds [%.1f MB/s]" % (size_mb, int(trans_time), rate))
+
+            # Update the FileCopy (if exists), or insert a new FileCopy
+            try:
+                done = False
+                while not done:
+                    try:
+                        fcopy = ar.ArchiveFileCopy\
+                                  .select()\
+                                  .where(ar.ArchiveFileCopy.file == req.file,
+                                         ar.ArchiveFileCopy.node == node)\
+                                  .get()
+                        fcopy.has_file = 'Y'
+                        fcopy.wants_file = 'Y'
+                        fcopy.save()
+                        done = True
+                    except pw.OperationalError:
+                        log.error("MySQL connexion dropped. Will attempt to reconnect in "
+                                  "five seconds.")
+                        time.sleep(5)
+                        db.config_connect()
```

Comment:
I agree that this isn't great. It's not nearly as necessary since we started using the peewee RetryOperation mixin, or what ever it's called. Now that we've customised that we could implement slightly more complex logic too, with delays and more attempts to restart. I'm going to leave this for the moment, but add changing it to our todo list.